### ISSUE TITLE
Upgrade to Spring Boot 1.5.11 and Spring Boot 2.0.1

### ISSUE THREAD

**Issue by jdubois** [jhipster/generator-jhipster#7422]
We need to upgrade our JHipster 4 maintenance branch to Spring Boot 1.5.11, and our JHipster 5 branch to Spring Boot 2.0.1, as they correct some very important security issues:

https://spring.io/blog/2018/04/05/multiple-cve-reports-published-for-the-spring-framework

This is a top priority for me, and I wanted to do it tonight, however, and as always, the official Spring BOM is not released yet!!!
To be more precise, I'm speaking of this artefact: http://repo1.maven.org/maven2/io/spring/platform/platform-bom/
By the way, we're still waiting for the official BOM for Spring Boot 2 (it's still a SNAPSHOT at the moment), and this is one of the main reasons why we can have a JHipster 5 stable release.

-> the more this goes, the more it seems to me we did a major error to rely on the Spring official BOM, as it doesn't seem to be updated at the speed we need

Should we just do our own BOM and replace the Spring one? That would be a big job, but that's what we used to do in the past, and it was working much better.

-> we need to decide this very fast, as we can't stay with the unpatched security issues

**Comment by jdubois** [comment]
For JHipster v5 this should already be corrected with https://github.com/jhipster/jhipster/commit/cab69f3f1ce4d29a70e376218d66d86b4f643a6d as we use SNAPSHOT releases, so every JHipster 5 project will automatically download that new version.

(please note that with Gradle you need to do `./gradlew build --refresh-dependencies`)

**Comment by jdubois** [comment]
For JHipster 4 we use a real release, which I have pushed to Maven Central (see https://github.com/jhipster/jhipster-dependencies/commit/99aa992eac30dae6a962f564a81e15aaad1eeb92 ). As the BOM isn't ready, I forced all Spring Boot dependencies to use 1.5.1.RELEASE.

This can take up to 2 hours to be synchronized, once it's done I will push the fix on jhipster/generator-jhipster on our v4_maintenance branch, and I will close this. I'll then do a maintenance release.

**Comment by jdubois** [comment]
For JHipster 4 this is done in https://github.com/jhipster/generator-jhipster/commit/ec145bab3356ddeadef72952f8fb819b02c0a580 and I'm doing a maintenance release after some tests

**Comment by jdubois** [comment]
Travis & Docker Hub are taking forever to run, so I won't do the v4_maintenance release tonight.

Besides, it looks like there are some failing tests I need to fix first, so that will be released tomorrow.

**Comment by wilkinsona** [comment]
The Platform bom has always lagged behind Boot to varying degrees and it always will do. More so for new versions (like Cairo) where Boot and several other projects all need to be GA before the Platform can GA.

FWIW, Cairo should released early next week. However, if tracking Boot releases as closely as possible is a priority, then I agree that using the Platform bom was a mistake.

**Comment by jdubois** [comment]
Thanks @wilkinsona - sorry for the harsh words, it's just that we have 3 major security issues and with the BOM we are stuck, so I'm a bit under pressure. But indeed, I think we are more following Spring Boot that the platform BOM, and in that way the BOM might not be a good choice for us.

**Comment by jdubois** [comment]
So very bad news: Spring Boot 2.0.1 make all our tests fail, so I need to revert.... And it looks the same with Spring Boot 1.5.11...
I have no idea what has changed, but that's pretty bad.

So we are back to the original issue :-(

**Comment by philwebb** [comment]
Can you just bump the framework version? The CVEs aren't actually in Boot itself.

**Comment by jdubois** [comment]
As everything was failing:
- JHipster v5 is back to Spring Boot 2.0.0 (as we're on a SNAPSHOT it's important to be stable)
- JHipster v4 is still with Spring Boot 1.5.11 but hasn't been released
- I also needed to upgrade the JHipster Registry (which had the same issue), so it's back to the older Spring Boot 1.5.x and 2.0.0 releases so it's not failing anymore

I'll do more tests tomorrow, including bumping the framework version as @philwebb told (thanks Phil!). What a weird issue!!!

**Comment by philwebb** [comment]
@jdubois What's the actual failure? Got logs you can point me at?

**Comment by tianlinzx** [comment]
Spring Boot 2.0.1 and 1.5.11 are finally released.
https://spring.io/blog/2018/04/05/spring-boot-2-0-1-available-now

**Comment by deepu105** [comment]
Once this is sorted out we really need to find a solution for the BOM issue else it will slow us down

**Comment by jdubois** [comment]
@deepu105  the current "solution" is that we use the BOM and override all Spring Boot releases: it looks pretty bad, but in the end it works quite well, so we could stay like that for some time and see if that's OK

And more bad news: Docker Hub looks completely down, my Docker builds from yesterday are still queued, so even if I find a solution quickly we won't have Docker builds ready :-(

**Comment by erikkemperman** [comment]
@jdubois We could go back to inheriting from `spring-boot-dependencies` as I originally intended. We don't use that many additional dependencies, actually, in the generator. But some of them might be convenient for end-users nevertheless to have as managed dependencies, hard to say...

@DanielFran I believe you were the one who requested we use `spring-platform`: this is introducing some delays in our ability to track latest spring-boot releases easily... Would you mind if it were dropped  and back to `spring-boot-dependencies` directly?

**Comment by jdubois** [comment]
Oh in fact I just realize the error we have!! As we only upgrade Spring Boot on top of `spring-platform`, we are using the newer Spring Boot release with the older Spring Framework version!
So typically on my test project I have Spring Boot 2.0.1 with Spring Framework 5.0.4:

- So it doesn't solve the security errors at all, as they are in Spring Framework and not Spring Boot (that's pretty bad!)
- We have inconsistent libraries, and that's probably the cause of our test failures

**Comment by erikkemperman** [comment]
@jdubois 
Not quite, I think: since `spring-platform` is a _child_ of `spring-boot-starter-parent`,
https://github.com/spring-io/platform/blob/vCairo-RC1/platform-bom/pom.xml#L9
, it's not enough to simply override the property like you tried yesterday:
https://github.com/jhipster/jhipster/commit/cab69f3f1ce4d29a70e376218d66d86b4f643a6d#diff-78a0a62b9524ac5cc8a07f55a4a29683R78

But as the security issues are not in boot but in the framework, you _could_ instead add
```xml
<spring.version>5.0.5.RELEASE</spring.version>
```
to override the version of the framework we receive.

Hope this helps.

For what it's worth, I have the impression that `spring-platform` tends to be "delayed" relative to spring-boot releases by an entirely reasonable amount, given that it also tracks lots of other upstream dependencies for us and has to make sure those work well together (which is a big benefit for end-users even if we don't use many of them in the generated jhipster projects out of the box).

**Comment by jdubois** [comment]
I see @wilkinsona just released a new platform BOM for Spring Boot 1.5.1: thank you **so much** Andy!! 👍 
I'm going to test this, and after many hours of hard work I'd rather stay on the platform BOM, it helps on so much dependency headaches!

**Comment by erikkemperman** [comment]
@jdubois Are you planning on starting a separate branch of jhipster-dependencies to support JH 4 / SB 1.5?

I ask because I was thinking to get rid of some dependencies that are no longer needed in JH5 but would have to remain to support JH4.

But you can't have a single release of jhipster-dependencies for both SB 1.5 and 2 so it'll have to be split anyway, right?

**Comment by erikkemperman** [comment]
@jdubois That should work, I guess. Although if it is expected that a lot of commits (like minor dep version bumps) should be merged "back" across repos it might become annoying -- can't simply cherry-pick because of the different path prefix! At some point we might want to have a maintenance branch on the jhipster repository?

**Comment by jdubois** [comment]
I think it's good for JHipster 4 - I'll probably do a release this afternoon.
For JHipster 5 I'm waiting for the Spring platform BOM "Cairo.RC2" (or maybe even RELEASE????), and I see @wilkinsona is working on it - so it's better to wait than do anything else. Anyway, that's not our released version (it's still in BETA), so it's not as important as the v4 branch.

@erikkemperman we'll do maintenance branches on jhipster/jhipster in the future - I know here we are losing a few commits, but that's not important, the v4 is in maintenance mode anyway, there won't be anything big happening there.

**Comment by jdubois** [comment]
@wilkinsona did you just released Cairo.RELEASE? I see it at http://repo1.maven.org/maven2/io/spring/platform/platform-bom/Cairo-RELEASE/ !!!
I'm upgrading right now

**Comment by erikkemperman** [comment]
Excellent!

**Comment by wilkinsona** [comment]
[I did](https://spring.io/blog/2018/04/09/spring-io-platform-cairo-release), but I also [announced the end of the Plaform's life](https://spring.io/blog/2018/04/09/spring-io-platform-end-of-life-announcement) at the same time. Cairo (and Brussels) will be maintained for 12 months to give everyone time to migrate.

**Comment by erikkemperman** [comment]
All right, so we'll have to decide when to migrate to use `spring-boot-dependencies` as the parent POM. I would propose sticking with Cairo for now. Thanks @wilkinsona!

**Comment by erikkemperman** [comment]
@jdubois Actually it's not that much work -- at least not to get the parent change working. We don't use many of the additional managed dependencies in an out-of-the-box generated app, I believe. However, users migrating will perhaps find that some of the previously managed dependencies need to now be done by themselves.

I don't think we should have the ambition to take over spring-platform scope, to be honest. Users will naturally discover what to do after upgrading. It'll be annoying, to be sure, but not difficult.

Does that make sense?

I can make a quick PR today.

**Comment by jdubois** [comment]
Yes totally agree @erikkemperman !!! In that case I'm waiting for your PR! But as Spring Cloud is still in milestone, can we have a stable release?

**Comment by jdubois** [comment]
just tested - Spring Platform CAIRO-RELEASE does depends on milestone releases, so you need to have access to the Spring Milestone Maven repo.

**Comment by erikkemperman** [comment]
Yes, I noticed the same thing, we need the spring milestones repos for the moment. It looks like spring-cloud should have RC1 next week and RELEASE by the end of the month. 

Do you want to wait with JH v5 release until then?

But at least spring-cloud-stream has just released Elmhurst GA (I'll upgrade in the same PR) so we only have to wait for:
https://github.com/spring-cloud/spring-cloud-release/milestones

**Comment by jdubois** [comment]
No, that's too much waiting - let's keep the milestone repos, and if everything is fine we'll release very soon

**Comment by erikkemperman** [comment]
I have it working -- mostly -- but the couchbase sample for reasons I don't yet understand. I have to leave it here for today, but will make a PR in case someone else wants to try.

**Comment by tchlyah** [comment]
Yes sure, let me check

**Comment by erikkemperman** [comment]
@tchlyah I think we may only have to upgrade `couchbase-testcontainer` to 1.12 (to accomodate couchbase-java-client 2.5.7). I'm trying it now.

**Comment by erikkemperman** [comment]
See also https://github.com/differentway/testcontainers-java-module-couchbase/commit/23db92fce383a002f43cd4ff744df893a624c0a0

**Comment by tchlyah** [comment]
Yes, I already done a pull request on your branch which fix the issue. You should just merger it.

**Comment by erikkemperman** [comment]
Ah! Thanks! Strange, I didn't get an email about that... Anyway, will merge it and Travis should go green :-)

**Comment by cbornet** [comment]
BTW it’s now Spring Boot 1.5.12 : https://spring.io/blog/2018/04/10/spring-boot-1-5-12-available-now

**Comment by jdubois** [comment]
Yes thanks @cbornet ! But as of this morning no new Cairo release :-)
Indeed I now understand we need to move, and @erikkemperman did a great job here, it's nearly done.

I'm mostly sad for people using JHipster 4, as we won't be able to maintain it very long.

**Comment by wilkinsona** [comment]
There's no need for a new Cairo release. It was only Spring Framework 4.3.x and, therefore Spring Boot 1.5.x, that was affected by the incomplete fix for the security vulnerability. A new IO Platform Brussels release will be published later today.

**Comment by jdubois** [comment]
Oh sorry @wilkinsona - I meant Brussels, yes! And we'll release after you, of course!

**Comment by jdubois** [comment]
Closing as:

- For JHipster 4 I have upgraded to Brussels-SR9
- For JHipster 5 we are now directly using the Spring Boot dependencies BOM, see https://github.com/jhipster/jhipster/pull/55
